doop-api/metrics.go

```diff
@@ -0,0 +1,201 @@
+		),
+		groupedViolationsGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "doop_grouped_violations",
+				Help: "Number of violation groups, grouped by constraint, source cluster and selected object identity labels.",
+			},
+			append([]string{"template_kind", "constraint_name"}, objectIdentityLabels...),
+		),
+		auditAgeOldestGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "doop_oldest_audit_age_seconds",
+				Help: "Data age for each source cluster.",
+			},
+			[]string{"cluster"},
+		),
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (mc *MetricCollector) Describe(ch chan<- *prometheus.Desc) {
+	mc.rawViolationsGauge.Describe(ch)
+	mc.groupedViolationsGauge.Describe(ch)
+	mc.auditAgeOldestGauge.Describe(ch)
+}
+
+// Collect implements the prometheus.Collector interface.
+func (mc *MetricCollector) Collect(ch chan<- prometheus.Metric) {
+	descCh := make(chan *prometheus.Desc, 1)
+	mc.rawViolationsGauge.Describe(descCh)
+	rawViolationsDesc := <-descCh
+	mc.groupedViolationsGauge.Describe(descCh)
+	groupedViolationsDesc := <-descCh
+	mc.auditAgeOldestGauge.Describe(descCh)
+	auditAgeOldestDesc := <-descCh
+
+	//using the individual reports, we can immediately calculate the audit age
+	reports, err := mc.downloader.GetReports()
+	if err != nil {
+		logg.Error("could not download reports for metric computation: %s", err.Error())
+	}
+	for clusterName, report := range reports {
+		ch <- prometheus.MustNewConstMetric(
+			auditAgeOldestDesc,
+			prometheus.GaugeValue, float64(oldestAuditTimestampForClusterReport(clusterName, report)),
+			clusterName,
+		)
+	}
+
+	//counting violation groups requires an aggregated report
+	fullReport := AggregateReports(reports, BuildFilterSet(url.Values{}))
+	for _, rt := range fullReport.Templates {
+		for _, rc := range rt.Constraints {
+			countViolationsForConstraint(rt.Kind, rc, mc.objectIdentityKeys, rawViolationsDesc, groupedViolationsDesc, ch)
+		}
+	}
+}
+
+func oldestAuditTimestampForClusterReport(clusterName string, report doop.Report) int64 {
+	result := int64(-1)
+	for _, rt := range report.Templates {
+		for _, rc := range rt.Constraints {
+			auditTimeStr := rc.Metadata.AuditTimestamp
+			if auditTimeStr == "" {
+				continue
+			}
+
+			auditTime, err := time.Parse(time.RFC3339, auditTimeStr)
+			if err != nil {
+				logg.Error("cannot parse audit timestamp %q for cluster %s: %s",
+					auditTimeStr, clusterName, err.Error())
+				//to ensure that the error is noticed, report a very old timestamp that likely triggers alerts
+				return 0
+			}
+
+			unixTime := auditTime.Unix()
+			if result == -1 || result > unixTime {
+				result = unixTime
+			}
+		}
+	}
+	return result
+}
+
+func countViolationsForConstraint(templateKind string, rc doop.ReportForConstraint, oidKeys []string, rawViolationsDesc, groupedViolationsDesc *prometheus.Desc, ch chan<- prometheus.Metric) {
```

Comment:
I don't really see how.

- `rawViolationsDesc`, `groupedViolationsDesc` and `ch` are temporary state for this particular scrape, so we cannot put them in the MetricCollector struct.
- `templateKind` and `rc` arise from the loop iterations at the callsite.
- `oidKeys` could be retrieved from the MetricCollector if this was a method instead of a function. I considered making it so, but did not find the change all that compelling for only one argument less.

In the end, I decided to leave this as-is because most arguments have very distinctive types (or, in other words, it's still way better than a `func(string, string, bool, int, bool)`.